Mortar constraint variable initialization for boundary materials

recuero opened the pull request:
The goal of this PR is to allow Mortar constraints and materials/aux kernels to interact through the coupling of variables and material properties. The code design strategy is to reinitialize objects on boundaries after having reinitialized lower dimensional variables, as described [here](https://github.com/idaholab/moose/pull/15086#issuecomment-624237395). 

This PR includes augmented API for coupling lower-d variables in #15367 
Discussion and design #15086 
Refs #14821

recuero committed:
Initial stab at reiniting Mortar variables for use in Materials Refs. #14821 #15086

recuero committed:
Initialization for ElemAuxBCs

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/522007/) on 54e8256 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15259/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15259/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 034dd023a38374d38a6dcf0742e83e1b8ebc5fd6`

recuero commented:
@lindsayad as the code stands, it prevents the segmentation fault posted [here](https://github.com/idaholab/moose/pull/15086#issuecomment-616884682) but the Mortar variable interrogated at the quadrature point in `Material::computeQpProperties` is zero at all times. This issue appears still to be linked to `Material` (or its initialization) because a different Mortar constraint in the same system can access the same lower-d variable appropriately.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/522983/) on b38022e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15259/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15259/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 034dd023a38374d38a6dcf0742e83e1b8ebc5fd6`

recuero committed:
Reinit lower-d variable for Jacobian and UserObject threads

recuero committed:
Adapt MooseMesh to new definition of libmesh has_interior_parent()

recuero committed:
Reinitialize always lower-d elements in displaced mesh

recuero committed:
Adapt MooseMesh to new interior_element API

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/527007/) on 4311bae wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15259/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15259/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 741715996179fc2e4bbb7d80175a44cd9d5f8dd0`

recuero committed:
Enforce mechanical contact Lagrange multiplier must be the same order as the mesh in NormalNodalLMMechanicalContact

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/530091/) on 0df4eb1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15259/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15259/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 741715996179fc2e4bbb7d80175a44cd9d5f8dd0`

lindsayad committed:
Add some coupledValueLower APIs

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/538432/) on 3080103 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15259/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15259/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0289ba37e81d5f13428700d907d1a00441d1c9c7`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/538459/) on 2fda30a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15259/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15259/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0289ba37e81d5f13428700d907d1a00441d1c9c7`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/538498/) on 191d519 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15259/site)

This comment will be updated on new commits.

recuero committed:
Remove error if an interior parent's side is invalid.

Remove debugging statements.

recuero committed:
Add test exercising coupled vars between aux kernel and mortar constraint

recuero committed:
Guard reinitLowerDElem against problems with no displaced mesh

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -1737,6 +1737,30 @@ FEProblemBase::reinitElemFace(const Elem * elem,
+{
+  // - Set our _current_lower_d_elem for proper dof index getting in the moose variables
+  // - Reinitialize all of our lower-d FE objects so we have current phi, dphi, etc. data
+  // _assembly(tid).reinitLowerDElemRef(elem, pts, weights);
+  _assembly[tid]->reinitMortarElem(lowerDElem); // Is it?
```
You'll still want to override `reinitLowerDElem` in `FEProblemBase` to look like this:
```c++
SubProblem::reinitLowerDElem(lower_d_elem, pts, weights, tid);

if (_displaced_problem)
  _displaced_problem->reinitLowerDElem(_displaced_mesh->elemPtr(lower_d_elem->id()), pts, weights, tid);
```
Note the `_displaced_mesh->elemPtr` code. That's critical. Otherwise you may be trying to reinit an element that doesn't even exist on the displaced mesh.

In the end product, you should have no new code in `DisplacedProblem`. Just the base implementation in `SubProblem` and an override in `FEProblemBase`

recuero replied:
@lindsayad thanks for the review!

Tried to follow your design suggestions. One of the tests is failing: `mortar/continuity-2d-conforming.equalgradient`. The issue appears to come from the `fe_lower->reinit(elem);` call in `Assembly::reinitLowerDElem`. An assertion in libmesh in `FE::reinit` is acting out, `libmesh_assert(this->qrule);`. Could it be that there is no default quadrature rule for that problem?

<details>
  <summary>Backtrace</summary>
1: 1   libmesh_dbg.0.dylib                 0x00000001077ebb68 libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*) + 232
2: 2   libmesh_dbg.0.dylib                 0x00000001078ceadb **libMesh::FE<1u, (libMesh::FEFamily)0>::reinit**(libMesh::Elem const*, std::__1::vector<libMesh::Point, std::__1::allocator<libMesh::Point> > const*, std::__1::vector<double, std::__1::allocator<double> > const*) + 427
3: 3   libmoose-dbg.0.dylib                0x0000000102a7f34a **Assembly::reinitLowerDElem**(libMesh::Elem const*, std::__1::vector<libMesh::Point, std::__1::allocator<libMesh::Point> > const*, std::__1::vector<double, std::__1::allocator<double> > const*) + 394
4: 4   libmoose-dbg.0.dylib                0x0000000101e57a26 **SubProblem::reinitLowerDElem**(libMesh::Elem const*, unsigned int, std::__1::vector<libMesh::Point, std::__1::allocator<libMesh::Point> > const*, std::__1::vector<double, std::__1::allocator<double> > const*) + 70
5: 5   libmoose-dbg.0.dylib                0x0000000101e57935 **FEProblemBase::reinitLowerDElem**(libMesh::Elem const*, unsigned int, std::__1::vector<libMesh::Point, std::__1::allocator<libMesh::Point> > const*, std::__1::vector<double, std::__1::allocator<double> > const*) + 69
6: 6   libmoose-dbg.0.dylib                0x00000001024cb352 **ComputeResidualThread::onBoundary(**libMesh::Elem const*, unsigned int, short, libMesh::Elem const*) + 290
7: 7   libmoose-dbg.0.dylib                0x0000000101d258ee ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator()(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool) + 990
8: 8   libmoose-dbg.0.dylib                0x0000000102588dec void libMesh::Threads::parallel_reduce<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, ComputeResidualThread>(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, ComputeResidualThread&) + 156
9: 9   libmoose-dbg.0.dylib                0x0000000102581a9f NonlinearSystemBase::computeResidualInternal(std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 863
10: 10  libmoose-dbg.0.dylib                0x00000001025811e3 NonlinearSystemBase::computeResidualTags(std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 627
11: 11  libmoose-dbg.0.dylib                0x0000000101e7f71a FEProblemBase::computeResidualTags(std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 2346
12: 12  libmoose-dbg.0.dylib                0x0000000101e7e866 FEProblemBase::computeResidualInternal(libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&, std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 230
13: 13  libmoose-dbg.0.dylib                0x0000000101e7e4cc FEProblemBase::computeResidual(libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&) + 252

</details>

recuero replied:
```      
      QGauss qrule(elem->dim(), fe_type.default_quadrature_order());
      fe_lower->attach_quadrature_rule(&qrule);
      fe_lower->reinit(elem);
```

Maybe this will do the trick.

lindsayad commented on modules/contact/include/constraints/NormalNodalLMMechanicalContact.h:
```diff
@@ -38,4 +41,8 @@ class NormalNodalLMMechanicalContact : public NodeFaceConstraint
 
   const MooseEnum _ncp_type;
+
+  FEProblem & _fe_problem;
```
This should be able to be a `const` reference

lindsayad reviewed (changes requested):
Looking really good @recuero . For a final review we'll have to use someone else as I have commits on this PR

recuero committed:
Rename variables. Don't use camel notation for variables.

recuero committed:
Address comments: From reinitLowerDElemRef to reinitLowerDElem

recuero committed:
Attach quadrature rule when fe_lower->reinit does not have quadrature points.

lindsayad commented:
Looks like some interesting stuff going on with commits right now. Maybe just rebase what you had before on next? All the interface kernel stuff is now in next

recuero commented:
`next` was giving some errors too with private apps --I think. The commit mess should be fixed now.

recuero committed:
Fix bad rebase

lindsayad committed:
Fix quadrature rule (hopefully)

lindsayad commented on framework/include/loops/ProjectMaterialProperties.h:
```diff
@@ -39,7 +39,10 @@ class ProjectMaterialProperties : public ThreadedElementLoop<ConstElemPointerRan
+  virtual void onBoundary(const Elem * elem,
+                          unsigned int side,
+                          BoundaryID bnd_id,
+                          const Elem * /* lower_d_elem = nullptr*/) override;
```
defaults aren't actually inherited from base classes. I would uncomment the entire thing here. Only comment out in the .C implementation if it's unused

lindsayad commented on framework/include/loops/ThreadedElementLoopBase.h:
```diff
@@ -293,15 +302,17 @@ template <typename RangeType>
                                                 unsigned int /*side*/,
-                                                BoundaryID /*bnd_id*/)
+                                                BoundaryID /*bnd_id*/,
+                                                const Elem * /* nullptr*/)
```
The other commented code in this signature are names instead of values. I would say either do nullptr -> lower_d_elem or nullptr -> `lower_d_elem = nullptr`

lindsayad commented on framework/include/loops/ThreadedElementLoopBase.h:
```diff
@@ -293,15 +302,17 @@ template <typename RangeType>
                                                unsigned int /*side*/,
-                                               BoundaryID /*bnd_id*/)
+                                               BoundaryID /*bnd_id*/,
+                                               const Elem * /*=nullptr*/)
```
The other commented code in this signature are names instead of values. I would say either do nullptr -> lower_d_elem or nullptr -> `lower_d_elem = nullptr`

lindsayad commented on framework/src/loops/ComputeMarkerThread.C:
```diff
@@ -81,7 +81,10 @@ ComputeMarkerThread::onElement(const Elem * elem)
+ComputeMarkerThread::onBoundary(const Elem * /*elem*/,
+                                unsigned int /*side*/,
+                                BoundaryID /*bnd_id*/,
+                                const Elem * /*=nullptr*/)
```
Per before let's make sure we also have a meaningful variable name in the comment in case an implement wants to come along and make use of it (although yes they should inspect the header :-))

lindsayad commented on modules/contact/include/constraints/NormalNodalLMMechanicalContact.h:
```diff
@@ -38,4 +41,8 @@ class NormalNodalLMMechanicalContact : public NodeFaceConstraint
+
+  const FEProblem & _fe_problem;
+
+  const DisplacedProblem * _displaced_problem;
```
Nice `const`!

lindsayad commented on test/tests/mortar/displaced-gap-conductance-2d-auxkernel/tests:
```diff
@@ -0,0 +1,13 @@
+    custom_cmp = 'auxkernel.cmp'
+    mesh_mode = 'REPLICATED'
+    max_parallel = 2
+    requirement = 'We shall be able to couple variables between auxiliary kernels and ad mortar constraints'
```
Remove the "ad" portion of this requirement, and add a period at the end of the sentence

lindsayad replied:
Also maybe say "boundary auxiliary kernels" or do we actually also support block aux kernels with mortar constraints? I'd have to think about that one

recuero replied:
You are right. It should be "boundary auxiliary kernels" here. Not sure if block-wise aux kernels are supported with Mortar constraints. I'd say probably not since it has not been tested yet.

lindsayad commented on test/tests/mortar/displaced-gap-conductance-2d-auxkernel/tests:
```diff
@@ -0,0 +1,13 @@
+    exodiff = 'gap-conductance_out.e'
+    custom_cmp = 'auxkernel.cmp'
+    mesh_mode = 'REPLICATED'
+    max_parallel = 2
```
Any particular reason for the `max_parallel`?

recuero replied:
No. I'll remove it. I think I only added it because it is a very small test.

recuero replied:
I believe we should also test a boundary material with Mortar constraints. I think that, as long as the material properties aren't stateful, it should work. I'll add it.

lindsayad reviewed (dismissed):
Looking really good. I'm about to  push up two commits that add some more `const-ness`. A general question is do you feel like the one test covers all the capability that we've added here? It's possible the answer is yes; it's why I'm asking :smile:

lindsayad committed:
Update API const-ness

recuero committed:
Address reviewer's comments.

- Define defaults in child classes
- Explicitly state variable name in comment when it's not used
- Add test checking for boundary restricted material use in Mortar classes

lindsayad reviewed (commented):
This is awesome work @recuero. Pretty cool to have such MOOSEY development from a non-MOOSE team member.

This has my approval, but we'll need another reviewer to formally approve since I have commits on this PR.

fdkong commented:
I am working on

fdkong commented on framework/include/base/Assembly.h:
```diff
@@ -1809,6 +1819,15 @@ class Assembly
+  const QBase * _const_current_qrule_lower;
+  /// quadrature rule used on lower dimensional elements. This should always be
+  /// the same as the face qrule
+  QBase * _current_qrule_lower;
```
Do we have two variables that point to the same thing? The comments are the same.

lindsayad replied:
Oh I guess I should add doxygen that explains why I need both like the explanation I gave in https://github.com/idaholab/moose/pull/15259#discussion_r442375074

lindsayad replied:
@recuero I will add this doxygen and also do the method renaming. You don't have to do anything right now

lindsayad replied:
I added comments explaining why we have const and non-const versions

fdkong commented on framework/include/interfaces/Coupleable.h:
```diff
@@ -155,6 +164,15 @@ class Coupleable
+   * @return Reference to a ADVariableValue for the coupled variable
+   */
+  const ADVariableValue & adCoupledValueLower(const std::string & var_name,
+                                              unsigned int comp = 0) const;
```
@lindsayad Could we handle `lower` via `adCoupledValue` by passing a parameter? We are doing that for API: 

```
  virtual void onBoundary(const Elem * elem,
                           unsigned int side,
                           BoundaryID bnd_id,
                           const Elem * lower_d_elem = nullptr) override;
```

I think consistency is important.

fdkong commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -1367,6 +1375,10 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
   std::unordered_map<SubdomainID, std::set<BoundaryID>> _neighbor_subdomain_boundary_ids;
 
+  /// Holds a map from a high-order element side to its corresponding lower-d element
+  std::map<std::pair<const Elem *, unsigned short int>, const Elem *>
```
`unordered_map`?

fdkong commented on framework/src/base/Assembly.C:
```diff
@@ -618,6 +619,19 @@ Assembly::setFaceQRule(QBase * qrule, unsigned int dim)
+{
+  // The lower-dimensional quadrature rule matches the face quadrature rule
+  _const_current_qrule_face = _const_current_qrule_lower = _current_qrule_face =
+      _current_qrule_lower = qrule;
```
Could we use only one variable if they are the same?

fdkong commented on framework/src/base/Assembly.C:
```diff
@@ -2116,26 +2130,38 @@ Assembly::reinitNeighborFaceRef(const Elem * neighbor,
+
+    if (weights)
+      lower_rule->setWeights(*weights);
+
```
`weights` is a passed-in parameter. How do you make sure the sum of them is unity? You have a comment:

`// This also sets the quadrature weights to unity`

lindsayad replied:
We cannot do an a priori check here. The sum of the weights should be equal to the area of the reference element and the same quadrature rule may be being applied to several different element types. I think we just have to trust the user that if they're going to pass in weights (which is currently not done anywhere in the framework) that they know what they're doing

fdkong commented on modules/contact/src/constraints/NormalNodalLMMechanicalContact.C:
```diff
@@ -46,10 +46,22 @@ NormalNodalLMMechanicalContact::NormalNodalLMMechanicalContact(const InputParame
-    _ncp_type(getParam<MooseEnum>("ncp_function_type"))
-
+    _ncp_type(getParam<MooseEnum>("ncp_function_type")),
+    _fe_problem(*getCheckedPointerParam<FEProblem *>("_fe_problem", "Must be using FEProblem.")),
```
`FEProblem` --> `FEProblemBase`?

recuero replied:
Yeah, I think it should be `FEProblemBase`. It should be addressed now.

fdkong commented on test/src/constraints/GapHeatConductanceAuxKernel.C:
```diff
@@ -0,0 +1,77 @@
+      {
+        auto gap = (_phys_points_slave[_qp] - _phys_points_master[_qp]).norm();
+        heat_transfer_coeff =
+            (0.5 * (_slave_gap_conductance[_qp] + _master_gap_conductance[_qp])) / gap;
```
`gap` will be always nonzero?

recuero replied:
As the tests are set up `gap` is always nonzero. Added an `assert` in case another application or test uses these ad hoc classes.

fdkong commented on test/tests/mortar/displaced-gap-conductance-2d-bnd-coupling/tests:
```diff
@@ -0,0 +1,20 @@
+    input = 'gap-conductance-bnd-aux-kernel.i'
+    exodiff = 'gap-conductance-bnd-aux-kernel_out.e'
+    custom_cmp = 'exodiff.cmp'
+    mesh_mode = 'REPLICATED'
```
@lindsayad Mortar does not work with distributed mesh yet?

lindsayad replied:
No it does not because I either need to be able to have mesh generators request ghosting or I need to put capability for push/pull element data in the LowerDBlocksFromSidesetGenerator. It should not be hard to do but I have not had time yet

fdkong reviewed (changes requested):
It looks nice for me.  A few comments

recuero committed:
Address reviewer's comments

recuero commented:
Thanks @lindsayad for your help, really enjoyed this work!

lindsayad commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -898,16 +898,30 @@ MooseMesh::cacheInfo()
   for (const auto & elem : getMesh().element_ptr_range())
   {
     SubdomainID subdomain_id = elem->subdomain_id();
+    const Elem * ip_elem = elem->interior_parent();
+
+    if (ip_elem)
+    {
+      unsigned int ip_side = ip_elem->which_side_am_i(elem);
+
+      // For some grid sequencing tests: ip_side == libMesh::invalid_uint
+      if (ip_side != libMesh::invalid_uint)
+      {
+        auto pair = std::make_pair(ip_elem, ip_side);
+        _higher_d_elem_side_to_lower_d_elem.insert(
+            std::pair<std::pair<const Elem *, unsigned short int>, const Elem *>(pair, elem));
+      }
+    }
```
This is excellent @recuero. I was actually just wondering what additional expense constructing this mesh data structure incurred but you piggy-backed off an existing element loop. :+1:

lindsayad committed:
Add adCoupledLowerValue API test

This simple test couples a lower dimensional variable into a boundary
condition of a higher dimensional variable. The jacobian is not perfect
for this at this time because there is no machinery in ADIntegratedBC to
properly handle the offset of lower dimensional variables.

This commit also:
- Renames adCoupledValueLower to adCoupledLowerValue
- Adds some commentary in `Assembly` about why we have const and
  non-const versions of qrules

lindsayad commented:
@keniley1 this PR adds the capability to easily couple lower-dimensional variables into things like (higher-dimensional) boundary conditions. I think I recall that that is something you're interested in for things like charge build-up on walls, etc.

lindsayad commented:
@fdkong this should be ready for review again
